### PR TITLE
[bitnami/thanos] Add hash and time base partion for thanos store

### LINKED ISSUE #4417: [bitnami/thanos] add support for sharding in storegateway
**Which chart**:

bitnami/thanos

**Is your feature request related to a problem? Please describe.**

By default, the thanos-storegateway will advertise that it can serve all timeseries and all blocks in the object store (ie S3-bucket). As such, you can not scale the number of stores and register them all with thanos-query. Thanos-query will complain about multiple stores that advertise the same data with non-unique labels. The storegateway can be put behind a loadbalancer, such as a kubernetes Service (of type ClusterIP), but the fact that communication happens over long lived gRPC connections means that, most of the time, there is no guarantee that each storegateway replica will serve an equal amount of requests.

This is a known issue and [documented on the thanos website](https://thanos.io/tip/thanos/sharding.md/) and can be solved by introducing sharding (aka partitioning) in the storegateway. Sharding allows each replica of the storegateway to only serve some blocks but not others. A request that spans multiple blocks will then automatically be fanned out to multiple stores.

**Describe the solution you'd like**

Introduce the configuration of optional time-based or hash-based partitioning (each nth-block). Default the store to use a single partition. The storegateway could be converted to a deployment with an index suffix, like so:

|desc|deployment name|
|---|---|
|first partition|thanos-storegateway-0
|second partition|thanos-storegateway-1
|third partition|thanos-storegateway-2

Store could also remain a statefulset, although I am not sure why it is a statefulset at the moment.

**Introducing a partition suffix to the store statefulset/deployment will cause it to be re-created. This could be considered a backwards incompatible change.**

**Describe alternatives you've considered**

Use the thanos chart from banzai-cloud, which already supports sharding. The banzai chart seems less actively maintained.
https://github.com/banzaicloud/banzai-charts/blob/master/thanos/values.yaml#L174

**Additional context**

|desc|link|
|---|---|
| Official documentation | https://thanos.io/tip/thanos/sharding.md/
| CNCF slack discussion in #thanos | https://cloud-native.slack.com/archives/CK5RSSC10/p1604923594031700
| Example implementation | https://github.com/observatorium/deployments/blob/bf1304b0d7bce2ae3fefa80412bb358f9aa176fb/environments/openshift/manifests/observatorium-template.yaml#L1514-L1521

### PR DESCRIPTION
**Pull request by psandhu79**
**Description of the change**

This adds hash and time based partion for the storegateway as documented here
https://thanos.io/tip/components/store.md/#time-based-partitioning
https://thanos.io/tip/thanos/sharding.md/#relabelling
 

**Benefits**

This will allow the thanos query to talk to multiple thanos store gateways which contain the specified timeseries databases

**Applicable issues**

  - fixes #4417

**Checklist** 
- [X] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/).
- [X] Variables are documented in the README.md
- [X] Title of the PR starts with chart name (e.g. [bitnami/<name_of_the_chart>])

### CODE REVIEW

File: bitnami/thanos/values.yaml [hunk 1]
```diff
@@ -1778,6 +1778,16 @@ storegateway:
     ##
     # maxUnavailable: 1
 
+  ## Setting hashPartioning will create multiple store deployments based on the number of shards specified using the hashmod of the blocks
+  ## See https://thanos.io/tip/thanos/sharding.md/#relabelling
+  hashPartioning: {}
```
Comment (migruiz4) [inline bitnami/thanos/values.yaml:1783]: Although this works with this:
```
{{- if .Values.storegateway.hashPartioning }}
  {{- $shards = int .Values.storegateway.hashPartioning.shards }}
  {{- $hashPartioning = true }}
{{- else }}
  {{- $shards = len .Values.storegateway.timePartioning }}
  {{- $timePartioning = true }}
{{- end }}
```
I'd the `values.yaml` should be descriptive and include the allowed values.
Could you please modify the logic in the statefulset so the `values.yaml` can be set similar to this?
```
  hashPartioning:
    shards: ""
```

**Review by migruiz4 — CHANGES_REQUESTED** [review]
Thank you very much for your contribution!

I suggested some changes below, could you please take a look?

File: bitnami/thanos/values.yaml (hunk 1 above)
Comment (psandhu79) [reply bitnami/thanos/values.yaml]: Done

File: bitnami/thanos/templates/storegateway/statefulset.yaml [hunk 2]
```diff
@@ -83,45 +99,65 @@ spec:
...
+          {{- end }}
+          {{- if $timePartioning }}
+          {{- $partion := (slice $.Values.storegateway.timePartioning  $index) | first }}
+          {{- if $partion.max }}
+          - "--max-time={{ $partion.max }}"
+          {{- end}}
+          {{- if $partion.min }}
+          - "--min-time={{ $partion.min }}"
+          {{- end}}
+          {{- end }}
```
[the remark this answers is not in the document]
Comment (psandhu79) [reply bitnami/thanos/templates/storegateway/statefulset.yaml]: done

File: bitnami/thanos/Chart.yaml [hunk 3]
```diff
@@ -28,4 +28,4 @@ name: thanos
 sources:
   - https://github.com/bitnami/bitnami-docker-thanos
   - https://thanos.io
-version: 5.1.0
+version: 5.2.0
```
[the remark this answers is not in the document]
Comment (psandhu79) [reply bitnami/thanos/Chart.yaml]: Done

**Comment by mhulscher** [comment]
It seems that podAffinity and podAntiAffinity are unaffected by this PR. This means that, in case of anti-affinity, all pods from all shards will not be scheduled on the same node.

However, in my opinion, replica-1 from shard-0 should not have anti-affinity to replica-1 from shard-1, for example. Shouldn't we ensure that (anti-)affinity only applies to the same deployment?

**Comment by psandhu79** [comment]
> It seems that podAffinity and podAntiAffinity are unaffected by this PR. This means that, in case of anti-affinity, all pods from all shards will not be scheduled on the same node.
> 
> However, in my opinion, replica-1 from shard-0 should not have anti-affinity to replica-1 from shard-1, for example. Shouldn't we ensure that (anti-)affinity only applies to the same deployment?

I was looking at the monggb shard for inspiration but does not look like what you are asking for is implemented there. I guess i would need to change the 

app.kubernetes.io/component: storegateway 

to include app.kubernetes.io/component: storegateway-$index

then change the anti-affinity line include the index , but i guess then i will also need to change the service and hpa :'(

**Comment by mhulscher** [comment]
> > It seems that podAffinity and podAntiAffinity are unaffected by this PR. This means that, in case of anti-affinity, all pods from all shards will not be scheduled on the same node.
> > However, in my opinion, replica-1 from shard-0 should not have anti-affinity to replica-1 from shard-1, for example. Shouldn't we ensure that (anti-)affinity only applies to the same deployment?
> 
> I was looking at the monggb shard for inspiration but does not look like what you are asking for is implemented there. I guess i would need to change the
> 
> app.kubernetes.io/component: storegateway
> 
> to include app.kubernetes.io/component: storegateway-$index
> 
> then change the anti-affinity line include the index , but i guess then i will also need to change the service and hpa :'(

I was thinking out loud. I am not saying that the affinity rules should be changed, but I did want to start a discussion.

However, now that you mention it, shouldn't the service and HPA be changed, any way? AFAIK thanos-query should connect to each storegateway pod individually, because each shard/pod advocates the blocks that it can serve. If thanos-query connects to a single service that targets all shards from all replicas, thanos-query will not know which shard can serve which blocks.

The banzai chart uses a single store service of type ClusterIP with clusterIP set to None (a headless service), when this service is resolved it returns all the IP-addresses of all pods, https://github.com/banzaicloud/banzai-charts/blob/master/thanos/templates/store-service.yaml#L19

Right now, bitnami/thanos defaults to omitting the clusterIP, which causes it to be automatically populated with an IP by Kubernetes. This is not a headless service and when looked up, will only return the service IP. See

* https://github.com/bitnami/charts/blob/a0408924a50f28c0ae9bbabc4d29f5905e8b7158/bitnami/thanos/values.yaml#L1696
* https://github.com/bitnami/charts/blob/master/bitnami/thanos/templates/storegateway/service.yaml#L19

Instead of using a headless service, another option would  be to create a regular ClusterIP service per shard and to let thanos-query connect to each service's ClusterIP.

**Comment by psandhu79** [comment]
Ah in my tests I used the headless service option, provided by this chart.

**Comment by mhulscher** [comment]
Does it still work with multiple replicas per shard? Or does thanos query complain about duplicate label sets?

**Comment by psandhu79** [comment]
Tested with 5 shards and 2 replicas each, i do not see any errors regarding duplicate label sets in the thanos query or the thanos ruler where i have configured rules for SLOs

**Comment by migruiz4** [comment]
Hi @psandhu79,

I'm sorry for the late response.

After some internal discussion, we consider it would be safer for others users if the Sharded StoreGateway were an additional feature and not the default behavior.

That means:
- Instead of overwriting the `storegateway/statefulset.yaml`, this feature should be declared in a separated template, i.e `storegateway/statefulset-sharded.yaml`
- Should include a new value, i.e `storegateway.sharded.enabled` that controlled whether to use sharded statefulsets or not.
- Minor version instead of major.

Additionally, regarding affinities and anti-affinities as @mhulscher, I'd recommend having a look at this section of the MongoDB sharded statefulset, especially the `dataNodeLoopId` variable that contains the Shard ID:
https://github.com/bitnami/charts/blob/d4fd5eb610c16816f9f94d73ca9c71ac63ff0549/bitnami/mongodb-sharded/templates/shard/shard-data-statefulset.yaml#L42-L58

https://github.com/bitnami/charts/blob/d4fd5eb610c16816f9f94d73ca9c71ac63ff0549/bitnami/mongodb-sharded/values.yaml#L776-L785

**Pull request closed without merging**